surfman-chains/lib.rs

```diff
@@ -54,22 +54,64 @@ use surfman::SurfaceType;
 use surfman_chains_api::SwapChainAPI;
 use surfman_chains_api::SwapChainsAPI;
 
+/// An abstract swapchain backend that is responsible for managing surfaces
+/// used by that swapchain.
+pub trait SurfaceProvider {
+    /// A notification that any previous front buffer can be recycled.
+    fn recycle_front_buffer(&mut self, device: &mut Device, context_id: ContextID);
+
+    /// A notification to recycle the provided surface.
+    fn recycle_surface(&mut self, surface: Surface);
+
+    /// The swapchain requires a new surface.
+    fn provide_surface(
+        &mut self,
+        device: &mut Device,
+        context: &mut Context,
+        context_id: ContextID,
+        size: Size2D<i32>,
+    ) -> Result<Surface, Error>;
+
+    /// Obtain the current front buffer surface from this provider.
+    fn take_front_buffer(&mut self) -> Option<Surface>;
+
+    /// Inform the provider of the new front buffer surface.
+    fn set_front_buffer(
+        &mut self,
+        device: &mut Device,
+        context: &mut Context,
+        context_id: ContextID,
+        new_front_buffer: Surface,
+    ) -> Result<(), Error>;
+
+    /// Return a new surface with the desired size.
+    fn create_sized_surface(
+        &mut self,
+        device: &mut Device,
+        context: &mut Context,
+        size: Size2D<i32>,
+    ) -> Result<Surface, Error>;
+
+    /// Ensure any outstanding tracked surfaces are appropriately destroyed.
+    fn destroy_all_surfaces(
+        &mut self,
+        device: &mut Device,
+        context: &mut Context,
+    ) -> Result<(), Error>;
+}
+
 // The data stored for each swap chain.
 struct SwapChainData {
+    // The backend that provides surfaces for this swapchain.
+    surface_provider: Box<dyn SurfaceProvider + Send>,
```

Comment:
Can we avoid using a trait object here? E.g. make SwapChainData take a Provider type argument?

Reply:
I don't believe there is a way to enable that. The code that declares types for swapchains is kept apart from the code that knows where to pick up the provider types from.

Reply:
OK, we can revisit that when the Great API Refactoring happens.